src/glpk/env/alloc.c

```diff
@@ -45,74 +46,141 @@
 *  for book-keeping the memory usage statistics. */
 
 #ifdef HAVE_ENV
+
+/**
+ * "Unwrap" the memory block descriptor
+ */
+static inline MBD* ptr_to_mbd(void* ptr)
+{
+    MBD* mbd = (MBD *)((char *)ptr - MBD_SIZE);
+    xassert(mbd->self == mbd);
+    return mbd;
+}
+
+/**
+ * Take the mbd off the environment. Must only ever try to remove an mdb from the environment
+ * it was allocated on
+ */
+static inline void _remove_from_env(MBD* mbd, ENV* env)
+{
+    xassert(mbd->env == env);
+
+    MBD* before = mbd->prev;
+    MBD* after = mbd->next;
+
+    /* unlink ourselves */
+    if(after) {
+        after->prev = before;
+    }
+    if(before) {
+        before->next = after;
+    } else {
+        env->mem_ptr = after;
+    }
+
+    /* update counters */
+    if (get_mem_count() < 1 || get_mem_total() < mbd->size) {
+        xerror("unlinking mbd: memory allocation error\n");
+    }
+    add_mem_count(-1);
+    add_mem_total(-(mbd->size));
+}
+
+/**
+ * Free an mbd from an env. If the mbd's env and the current env mismatch, the block will *not* be deallocated yet,
+ * instead the deallocation will be deferred until the environment is freed
+ */
+static inline void* _free_mbd(MBD* mbd, ENV* env) {
```

Comment:
I'm curious about the use of 'inline'. I thought that most modern compilers were smart enough to figure out whether to inline or not automatically.

Reply:
It serves 2 main purposes 1) It informs the compiler of my intent for it to be inline, and so perhaps increases the odds of that actually happening. and 2) It communicates the intent to readers of this code that they should not take the address of it, as that would prevent it from being inlined. 

Incidentally, the compiler may still choose not to inline it. You have to jump through more (compiler-specific) hoops to force the compiler to inline it. (for example in gcc `static inline __attribute__((__always_inline__))` will  at least cause taking its address to generate a warning with `-Winline`)